### ISSUE TITLE
What is `${p}` clober in some instructions definition of ARMv8.1-M instruction tabelgen?

### ISSUE THREAD

**Issue by ChoKyuWon** [llvm/llvm-project#96490]
I found that the new ARMv8.1-M ISA instructions table gen generates weird code.
For example, the PACG definitions is like this:
```
let hasSideEffects = 1 in {
class PACBTIAut<dag iops, string asm, bit b>
  : V8_1MI<(outs), iops,
           AddrModeNone, NoItinerary, asm, "$Ra, $Rn, $Rm", "", []> {
  bits<4> Ra;
  bits<4> Rn;
  bits<4> Rm;
  let Inst{31-20} = 0b111110110101;
  let Inst{19-16} = Rn;
  let Inst{15-12} = Ra;
  let Inst{11-5}  = 0b1111000;
  let Inst{4}     = b;
  let Inst{3-0}   = Rm;
}
}

def t2AUTG  : PACBTIAut<(ins pred:$p, GPRnosp:$Ra, GPRnopc:$Rn, GPRnopc:$Rm),
                        "autg${p}", 0>;
```
And it omits the following emitter:
```C++
uint64_t ARMMCCodeEmitter::getBinaryCodeForInstr(const MCInst &MI,
    SmallVectorImpl<MCFixup> &Fixups,
    const MCSubtargetInfo &STI) const {
...
    case ARM::t2AUTG:
    case ARM::t2BXAUT: {
      // op: Ra
      op = getMachineOpValue(MI, MI.getOperand(2), Fixups, STI);
      op &= UINT64_C(15);
      op <<= 12;
      Value |= op;
      // op: Rn
      op = getMachineOpValue(MI, MI.getOperand(3), Fixups, STI);
      op &= UINT64_C(15);
      op <<= 16;
      Value |= op;
      // op: Rm
      op = getMachineOpValue(MI, MI.getOperand(4), Fixups, STI);
      op &= UINT64_C(15);
      Value |= op;
      break;
    }
```

So my question is: why do we need `${p}` for the first ins argument, and what are the first and second operands of MI?
Can we just remove all of {p} there and sort the operand order as normal instructions?

Here is the reference: https://developer.arm.com/documentation/ddi0553/latest/ and I cannot find any difference between `AUTG` and `MUL` in term of Instruction encoding, and want sure that is intended feature.

**Comment by ChoKyuWon** [comment]
@EugeneZelenko Can you change the tag to `backend:ARM`?

**Comment by DavidSpickett** [comment]
I think this is the condition code that most ARM instructions can have:
```
AUTG{<c>} <Ra>, <Rn>, <Rm>
```
The `{<c>}` bit - https://developer.arm.com/documentation/den0042/a/Unified-Assembly-Language-Instructions/Instruction-set-basics/Conditional-execution

```
PACBTIAut<(ins pred:$p, GPRnosp:$Ra, GPRnopc:$Rn, GPRnopc:$Rm)...
```
It may be 'p' because you might also call it a "predicate" (just a guess). Arm does not, as "predicate" normally refers to predication when using vectors.

@stuij correct me if I'm wrong here.

Also it's called `t2AUTG` despite the manual labelling it `T1` I think because `t2` refers to the thumb2 instruction set. Then `T1` as shown in the manual is the first encoding variant in that set. If you have a `T1` and a `T2` it's often 16 and 32 bit encodings.

**Comment by DavidSpickett** [comment]
>Here is the reference: https://developer.arm.com/documentation/ddi0553/latest/ and I cannot find any difference between t2AUTG and t2MUL in term of Instruction encoding, and want sure that is intended feature.

t2AUTG:
```
  let Inst{31-20} = 0b111110110101;
  let Inst{19-16} = Rn;
  let Inst{15-12} = Ra;
  let Inst{11-5}  = 0b1111000;
  let Inst{4}     = b;
  let Inst{3-0}   = Rm;
```

t2MUL:
```
  let Inst{31-27} = 0b11111;
  let Inst{26-23} = 0b0110;
  let Inst{22-20} = 0b000;
  let Inst{15-12} = 0b1111; // Ra = 0b1111 (no accumulate)
  let Inst{7-4} = 0b0000; // Multiply
```

Looks like 22-20 are different. This matches up with the manual (which can be tricky because it likes to list bits as 16-0 twice, instead of 31-0 once).

Perhaps you can show an example with `llvm-mc` where the encodings overlap?

**Comment by ChoKyuWon** [comment]
@DavidSpickett Thank you for your kind reply!

I mention the `t2MUL` because `MUL` instruction also requires the `<c>` assembler syntax according to the manual, but the tablegen did not implement it.
```
MUL<c>.W <Rd>, <Rn>{, <Rm>}
// Inside IT block, and <Rd>, <Rn>, <Rm> can be represented in T1
MUL{<c>}{<q>} <Rd>, <Rn>{, <Rm>}
```

Some instructions also have `<c>` or `<q>` assembler syntax, but It seems they are randomly implemented on the tablegen description.
I wonder why `t2PACG/AUTG` especially took the `pred` as its input, while `t2MUL` did not.

**Comment by DavidSpickett** [comment]
Ok I understand now, I assumed you meant the bit values not the way the assembler is specified (I used to work on the team implementing these things, I spent a lot of time staring at bit layouts).

> Some instructions also have <c> or <q> assembler syntax, but It seems they are randomly implemented on the tablegen description.

This is quite likely, we had similar problems with the `.w` suffix where it was implemented (until recently I think) on a case by case basis.

> I wonder why t2PACG/AUTG especially took the pred as its input, while t2MUL did not.

@stuij can you check if this is, as I suspect, an oversight, or whether there's a special reason?

@ChoKyuWon If you have noticed any other instructions, we can open an issue to track the general problem. It'll likely be low priority but if you're blocked on specific instructions do let us know that.
